File: python/graphs/d_star_lite_production.py
from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass
from enum import Enum
from heapq import heappop, heappush
from math import hypot, inf, sqrt


@dataclass(frozen=True, order=True)
class Cell:
    row: int
    col: int


class Heuristic(str, Enum):
    MANHATTAN = "manhattan"
    EUCLIDEAN = "euclidean"
    CHEBYSHEV = "chebyshev"
    OCTILE = "octile"


Key = tuple[float, float]


class DStarLite:
    def __init__(
        self,
        grid: Sequence[Sequence[bool]],
        start: Cell,
        goal: Cell,
        heuristic: Heuristic = Heuristic.MANHATTAN,
    ) -> None:
        if not grid or not grid[0]:
            raise ValueError("grid must be non-empty")

        self.rows = len(grid)
        self.cols = len(grid[0])
        if any(len(row) != self.cols for row in grid):
            raise ValueError("grid must be rectangular")

        self._validate_cell(start, "start")
        self._validate_cell(goal, "goal")

        self._start_id = self._id(start)
        self._goal_id = self._id(goal)
        self.heuristic = heuristic
        self.km = 0.0
        self.blocked = bytearray(self.rows * self.cols)
        for row, values in enumerate(grid):
            for col, value in enumerate(values):
                self.blocked[self._id(Cell(row, col))] = bool(value)

        if self.blocked[self._start_id] or self.blocked[self._goal_id]:
            raise ValueError("start and goal must be traversable")

        self.g = [inf] * self.cell_count
        self.rhs = [inf] * self.cell_count
        self.rhs[self._goal_id] = 0.0
        self._queue: list[tuple[float, float, int]] = []
        self._open_keys: list[Key | None] = [None] * self.cell_count
        self._push(self._goal_id)

    @property
    def cell_count(self) -> int:
        return self.rows * self.cols

    @property
    def start(self) -> Cell:
        return self._cell(self._start_id)

    @property
    def goal(self) -> Cell:
        return self._cell(self._goal_id)

    def replan(self) -> list[Cell]:
        self.compute_shortest_path()
        return self.extract_path()

    def set_blocked(self, cell: Cell, blocked: bool) -> bool:
        self._validate_cell(cell, "cell")
        cell_id = self._id(cell)
        if cell_id == self._start_id or cell_id == self._goal_id:
            raise ValueError("start and goal cannot be changed")
        if bool(self.blocked[cell_id]) == blocked:
            return False

        self.blocked[cell_id] = blocked
        if blocked:
            self.g[cell_id] = inf
            self.rhs[cell_id] = inf
            self._open_keys[cell_id] = None
        else:
            self._update_vertex(cell_id)

        for neighbor in self._adjacent(cell_id):
            if not self.blocked[neighbor]:
                self._update_vertex(neighbor)
        return True

    def move_start(self, new_start: Cell) -> None:
        self._validate_cell(new_start, "new start")
        new_start_id = self._id(new_start)
        if self.blocked[new_start_id]:
            raise ValueError("new start must be traversable")

        self.km += self._heuristic(self._start_id, new_start_id)
        self._start_id = new_start_id

    def compute_shortest_path(self) -> None:
        start_id = self._start_id
        while (
            self._key_less(self._top_key(), self._calculate_key(start_id))
            or self.rhs[start_id] != self.g[start_id]
        ):
            item = self._pop()
            if item is None:
                return

            old_key, current = item
            new_key = self._calculate_key(current)
            if self._key_less(old_key, new_key):
                self._push(current)
            elif self.g[current] > self.rhs[current]:
                self.g[current] = self.rhs[current]
                for predecessor in self._neighbors(current):
                    self._update_vertex(predecessor)
            else:
                self.g[current] = inf
                self._update_vertex(current)
                for predecessor in self._neighbors(current):
                    self._update_vertex(predecessor)

    def extract_path(self) -> list[Cell]:
        current = self._start_id
        if self.g[current] == inf:
            return []

        path = [self.start]
        for _ in range(self.cell_count):
            if current == self._goal_id:
                return path

            successors = self._neighbors(current)
            if not successors:
                return []

            current = min(successors, key=lambda node: (1.0 + self.g[node], node))
            if self.g[current] == inf:
                return []
            path.append(self._cell(current))

        return []

    def _update_vertex(self, cell_id: int) -> None:
        if self.blocked[cell_id]:
            self._open_keys[cell_id] = None
            return

        if cell_id != self._goal_id:
            self.rhs[cell_id] = min(
                (1.0 + self.g[neighbor] for neighbor in self._neighbors(cell_id)),
                default=inf,
            )

        self._open_keys[cell_id] = None
        if self.g[cell_id] != self.rhs[cell_id]:
            self._push(cell_id)

    def _push(self, cell_id: int) -> None:
        key = self._calculate_key(cell_id)
        self._open_keys[cell_id] = key
        heappush(self._queue, (*key, cell_id))

    def _top_key(self) -> Key:
        while self._queue:
            first, second, cell_id = self._queue[0]
            if self._open_keys[cell_id] == (first, second):
                return first, second
            heappop(self._queue)
        return inf, inf

    def _pop(self) -> tuple[Key, int] | None:
        while self._queue:
            first, second, cell_id = heappop(self._queue)
            if self._open_keys[cell_id] == (first, second):
                self._open_keys[cell_id] = None
                return (first, second), cell_id
        return None

    def _calculate_key(self, cell_id: int) -> Key:
        value = min(self.g[cell_id], self.rhs[cell_id])
        return value + self._heuristic(self._start_id, cell_id) + self.km, value

    def _heuristic(self, left_id: int, right_id: int) -> float:
        left = self._cell(left_id)
        right = self._cell(right_id)
        rows = abs(left.row - right.row)
        cols = abs(left.col - right.col)
        if self.heuristic is Heuristic.EUCLIDEAN:
            return hypot(rows, cols)
        if self.heuristic is Heuristic.CHEBYSHEV:
            return max(rows, cols)
        if self.heuristic is Heuristic.OCTILE:
            return max(rows, cols) + (sqrt(2) - 1) * min(rows, cols)
        return rows + cols

    def _neighbors(self, cell_id: int) -> list[int]:
        return [neighbor for neighbor in self._adjacent(cell_id) if not self.blocked[neighbor]]

    def _adjacent(self, cell_id: int) -> list[int]:
        row, col = divmod(cell_id, self.cols)
        result: list[int] = []
        if row > 0:
            result.append(cell_id - self.cols)
        if col + 1 < self.cols:
            result.append(cell_id + 1)
        if row + 1 < self.rows:
            result.append(cell_id + self.cols)
        if col > 0:
            result.append(cell_id - 1)
        return result

    def _id(self, cell: Cell) -> int:
        return cell.row * self.cols + cell.col

    def _cell(self, cell_id: int) -> Cell:
        return Cell(*divmod(cell_id, self.cols))

    def _validate_cell(self, cell: Cell, name: str) -> None:
        if not (0 <= cell.row < self.rows and 0 <= cell.col < self.cols):
            raise ValueError(f"{name} must be inside the grid")

    @staticmethod
    def _key_less(left: Key, right: Key) -> bool:
        return left < right

File: java/graphs/DStarLiteProduction.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.PriorityQueue;

public final class DStarLiteProduction {
    public record Cell(int row, int col) {}

    public enum Heuristic {
        MANHATTAN,
        EUCLIDEAN,
        CHEBYSHEV,
        OCTILE
    }

    private record Key(double first, double second) implements Comparable<Key> {
        @Override
        public int compareTo(Key other) {
            int firstComparison = Double.compare(first, other.first);
            return firstComparison != 0 ? firstComparison : Double.compare(second, other.second);
        }
    }

    private record QueueEntry(Key key, int id) implements Comparable<QueueEntry> {
        @Override
        public int compareTo(QueueEntry other) {
            int keyComparison = key.compareTo(other.key);
            return keyComparison != 0 ? keyComparison : Integer.compare(id, other.id);
        }
    }

    private final int rows;
    private final int cols;
    private final int cellCount;
    private final boolean[] blocked;
    private final double[] g;
    private final double[] rhs;
    private final Key[] openKeys;
    private final PriorityQueue<QueueEntry> open = new PriorityQueue<>();
    private final Cell goal;
    private final Heuristic heuristic;
    private Cell start;
    private double km;

    public DStarLiteProduction(boolean[][] walls, Cell start, Cell goal) {
        this(walls, start, goal, Heuristic.MANHATTAN);
    }

    public DStarLiteProduction(boolean[][] walls, Cell start, Cell goal, Heuristic heuristic) {
        validateGrid(walls);
        this.rows = walls.length;
        this.cols = walls[0].length;
        this.cellCount = Math.multiplyExact(rows, cols);
        this.start = Objects.requireNonNull(start, "start");
        this.goal = Objects.requireNonNull(goal, "goal");
        this.heuristic = Objects.requireNonNull(heuristic, "heuristic");
        validateCell(start, "start");
        validateCell(goal, "goal");

        blocked = new boolean[cellCount];
        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                blocked[id(new Cell(row, col))] = walls[row][col];
            }
        }
        if (isBlocked(id(start)) || isBlocked(id(goal))) {
            throw new IllegalArgumentException("start and goal must be traversable");
        }

        g = new double[cellCount];
        rhs = new double[cellCount];
        openKeys = new Key[cellCount];
        Arrays.fill(g, Double.POSITIVE_INFINITY);
        Arrays.fill(rhs, Double.POSITIVE_INFINITY);
        rhs[id(goal)] = 0.0;
        enqueue(id(goal));
    }

    public List<Cell> replan() {
        computeShortestPath();
        return extractPath();
    }

    public boolean setBlocked(Cell cell, boolean shouldBlock) {
        validateCell(cell, "cell");
        if (cell.equals(start) || cell.equals(goal)) {
            throw new IllegalArgumentException("start and goal cannot be changed");
        }

        int cellId = id(cell);
        if (isBlocked(cellId) == shouldBlock) {
            return false;
        }

        blocked[cellId] = shouldBlock;
        if (shouldBlock) {
            g[cellId] = Double.POSITIVE_INFINITY;
            rhs[cellId] = Double.POSITIVE_INFINITY;
            openKeys[cellId] = null;
        } else {
            updateVertex(cellId);
        }

        for (int neighbor : adjacent(cellId)) {
            if (!isBlocked(neighbor)) {
                updateVertex(neighbor);
            }
        }
        return true;
    }

    public void moveStart(Cell newStart) {
        validateCell(newStart, "new start");
        if (isBlocked(id(newStart))) {
            throw new IllegalArgumentException("new start must be traversable");
        }

        km += estimate(start, newStart);
        start = newStart;
    }

    public Cell start() {
        return start;
    }

    public Cell goal() {
        return goal;
    }

    private void computeShortestPath() {
        int startId = id(start);
        while (keyLess(topKey(), calculateKey(startId)) || rhs[startId] != g[startId]) {
            QueueEntry entry = pop();
            if (entry == null) {
                return;
            }

            Key newKey = calculateKey(entry.id());
            if (keyLess(entry.key(), newKey)) {
                enqueue(entry.id());
                continue;
            }

            if (g[entry.id()] > rhs[entry.id()]) {
                g[entry.id()] = rhs[entry.id()];
                for (int predecessor : neighbors(entry.id())) {
                    updateVertex(predecessor);
                }
                continue;
            }

            g[entry.id()] = Double.POSITIVE_INFINITY;
            updateVertex(entry.id());
            for (int predecessor : neighbors(entry.id())) {
                updateVertex(predecessor);
            }
        }
    }

    private List<Cell> extractPath() {
        int current = id(start);
        if (!Double.isFinite(g[current])) {
            return List.of();
        }

        List<Cell> path = new ArrayList<>(cellCount);
        path.add(start);
        for (int steps = 0; steps < cellCount; steps++) {
            if (current == id(goal)) {
                return List.copyOf(path);
            }

            int best = -1;
            for (int candidate : neighbors(current)) {
                if (best == -1 || g[candidate] < g[best]
                    || (g[candidate] == g[best] && candidate < best)) {
                    best = candidate;
                }
            }
            if (best == -1 || !Double.isFinite(g[best])) {
                return List.of();
            }

            current = best;
            path.add(cell(current));
        }
        return List.of();
    }

    private void updateVertex(int cellId) {
        if (isBlocked(cellId)) {
            openKeys[cellId] = null;
            return;
        }

        if (cellId != id(goal)) {
            double best = Double.POSITIVE_INFINITY;
            for (int successor : neighbors(cellId)) {
                best = Math.min(best, 1.0 + g[successor]);
            }
            rhs[cellId] = best;
        }

        openKeys[cellId] = null;
        if (g[cellId] != rhs[cellId]) {
            enqueue(cellId);
        }
    }

    private void enqueue(int cellId) {
        Key key = calculateKey(cellId);
        openKeys[cellId] = key;
        open.add(new QueueEntry(key, cellId));
    }

    private Key topKey() {
        while (!open.isEmpty()) {
            QueueEntry entry = open.peek();
            if (entry.key().equals(openKeys[entry.id()])) {
                return entry.key();
            }
            open.remove();
        }
        return new Key(Double.POSITIVE_INFINITY, Double.POSITIVE_INFINITY);
    }

    private QueueEntry pop() {
        while (!open.isEmpty()) {
            QueueEntry entry = open.remove();
            if (entry.key().equals(openKeys[entry.id()])) {
                openKeys[entry.id()] = null;
                return entry;
            }
        }
        return null;
    }

    private Key calculateKey(int cellId) {
        double value = Math.min(g[cellId], rhs[cellId]);
        return new Key(value + estimate(start, cell(cellId)) + km, value);
    }

    private double estimate(Cell left, Cell right) {
        int rowDistance = Math.abs(left.row() - right.row());
        int colDistance = Math.abs(left.col() - right.col());
        return switch (heuristic) {
            case EUCLIDEAN -> Math.hypot(rowDistance, colDistance);
            case CHEBYSHEV -> Math.max(rowDistance, colDistance);
            case OCTILE -> Math.max(rowDistance, colDistance)
                + (Math.sqrt(2.0) - 1.0) * Math.min(rowDistance, colDistance);
            case MANHATTAN -> rowDistance + colDistance;
        };
    }

    private List<Integer> adjacent(int cellId) {
        Cell current = cell(cellId);
        List<Integer> result = new ArrayList<>(4);
        if (current.row() > 0) {
            result.add(cellId - cols);
        }
        if (current.col() + 1 < cols) {
            result.add(cellId + 1);
        }
        if (current.row() + 1 < rows) {
            result.add(cellId + cols);
        }
        if (current.col() > 0) {
            result.add(cellId - 1);
        }
        return result;
    }

    private List<Integer> neighbors(int cellId) {
        List<Integer> result = new ArrayList<>(4);
        for (int neighbor : adjacent(cellId)) {
            if (!isBlocked(neighbor)) {
                result.add(neighbor);
            }
        }
        return result;
    }

    private boolean isBlocked(int cellId) {
        return blocked[cellId];
    }

    private int id(Cell cell) {
        return cell.row() * cols + cell.col();
    }

    private Cell cell(int cellId) {
        return new Cell(cellId / cols, cellId % cols);
    }

    private void validateCell(Cell cell, String name) {
        if (cell == null || cell.row() < 0 || cell.row() >= rows || cell.col() < 0 || cell.col() >= cols) {
            throw new IllegalArgumentException(name + " must be inside the grid");
        }
    }

    private static void validateGrid(boolean[][] walls) {
        if (walls == null || walls.length == 0 || walls[0] == null || walls[0].length == 0) {
            throw new IllegalArgumentException("grid must be non-empty");
        }
        int width = walls[0].length;
        for (boolean[] row : walls) {
            if (row == null || row.length != width) {
                throw new IllegalArgumentException("grid must be rectangular");
            }
        }
    }

    private static boolean keyLess(Key left, Key right) {
        return left.compareTo(right) < 0;
    }
}
